Allow also non-functions/classes to be included in module API summary

mhvk opened the pull request:
This is a possible solution to #117, in that it adds an option `:include-all-objects` to `automodapi` that allows one to make a summary of objects that are neither classes nor functions. It works by writing an `automodsumm` section with a new `:no-functions-or-classes` option.

I'm not sure about the approach, or what should be default, and would like to have comments before updating tests and documentation.

Cadair commented:
ooo this would be useful!

embray commented:
Thanks for tackling this.  I'm not so sure we need or want a special option for this though.  In the case of #117 for example it used to *just work*, but something changed in Sphinx that broke it, and I never tracked down what changed.

embray commented:
As you can see this also broke some tests, though it should be easy enough to fix.

embray commented on astropy_helpers/sphinx/ext/automodapi.py:
```diff
@@ -116,6 +121,15 @@
 """
 
+automod_templ_other = """
+Other variables
```
How about just "Global variables"?

mhvk commented:
> Thanks for tackling this. I'm not so sure we need or want a special option for this though. In the case of #117 for example it used to just work, but something changed in Sphinx that broke it, and I never tracked down what changed.

@embray - given the above, do you think I should pursue this PR, or is it better to try to find out what changed? If I should pursue it, is this the right direction? What should be the default?  (I'm slightly wary since I still don't understand the whole process all that well.)

embray commented:
I'm not entirely sure either. I didn't write any of this code.  My point is that if we're going to auto-generate documentation for a module that should include any module-level constants as well (at *least* if they have docstrings).  It should do this by default.

embray commented:
This section in the docs explains things: http://docs.astropy.org/en/stable/development/docguide.html#sphinx-extensions

embray commented on astropy_helpers/sphinx/ext/automodsumm.py:
```diff
@@ -119,6 +119,7 @@ class Automodsumm(BaseAutosummary):
     option_spec = dict(Autosummary.option_spec)
     option_spec['functions-only'] = flag
     option_spec['classes-only'] = flag
+    option_spec['no-functions-or-classes'] = flag
```
Maybe instead of `no-functions-or-classes`, which is awkward, this could just be called `'global-variables-only'`.

embray commented:
Now I'm *not* entirely sure if this actually used to work or not.  It's been a long time.  I think one thing that did break was display of docstrings for class attributes.  But that's different.

mhvk commented:
@embray - I agree that the only thing I now for sure used to work was class attributes, but those indeed go through a different code path. I like your suggestion of `global-variables-only` and think it makes most sense to display them by default *if* they have documentation *or* they are in `__all__`.

embray commented:
Sounds good.

eteq commented:
@mhvk @embray - :+1: on displaying constants if documented, but a clarification on this proposal: I think it probably is safer to *require* that they be in `__all__`.  My reasoning is that if you don't do that, then modules that do things like ``import numpy as np`` at the top without an `__all__` will be getting the numpy docs included. 

Another option might be to leave on (or maybe turn on?  I can't remember the default for sure...) the machinery that checks to see if the thing getting documented is actually a part of the same module.  But I'm guessing that's more confusing because it might hide stuff that a user actually wanted.

embray commented:
@eteq I'm definitely on board with only auto-documenting variables in `__all__`

mhvk commented:
I rebased this long-standing item -- still like the idea. I changed the names according to the suggestions, and hope to look at how to add tests after seeing what failures this gets (it failed in its initial incarnation).

mhvk commented:
actually, it looks like I should really make a PR to the new `sphinx-automodapi` package, right?

mhvk commented:
Since it seems this is now in the wrong package, see https://github.com/astropy/sphinx-automodapi/pull/24

If it indeed should be moved there, this one can be closed.

bsipocz commented:
@mhvk - Indeed, all the sphinx related extensions were moved already, and we only update the bundled version in ``extern`` to match newer ``sphinx-automodapi`` versions. Closing this now, as you've already opened the PR to the upstream repo.
